### ISSUE TITLE
ASGI opentelemetry.instrumentation packages incompatible with azure-monitor-opentelemetry-exporter

### ISSUE THREAD

**Issue by gautam-ergo** [Azure/azure-sdk-for-python#17839]
- **azure-monitor-opentelemetry-exporter**
- **Package Version - azure-monitor-opentelemetry-exporter = {version = "^1.0.0-beta.3", allow-prereleases = true}**
- **Operating System - macOS Mojave v10.14.6**
- **Python Version - v3.9.0**

**Describe the bug**
We are using FastAPI in our production and successfully implemented log exporting using Azure log handers to Azure App Insights.
On implementing OpenTelemetry,
 _**no versions of azure-monitor-opentelemetry-exporter match >1.0.0-beta.3,<2.0.0, opentelemetry-instrumentation-fastapi (>=0.19b0,<0.20) is incompatible with azure-monitor-opentelemetry-exporter (>=1.0.0-beta.3,<2.0.0)**_

ASGI open telemetry packages like,

[FastAPI Instrumentation](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation/opentelemetry-instrumentation-fastapi)
[Starlette Instrumentation](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation/opentelemetry-instrumentation-starlette)
[ASGI instrumentation](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation/opentelemetry-instrumentation-asgi)

face the same dependency conflict with [Azure monitor exporter](https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/monitor/azure-monitor-opentelemetry-exporter)

**To Reproduce**
After installing the required packages, try running the following FastAPI app:
`import os
import fastapi
import uvicorn
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchExportSpanProcessor
from azure.monitor.opentelemetry.exporter import AzureMonitorTraceExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor

app = fastapi.FastAPI()

@app.get("/foor")
async def foo():
    return {"message": "hello world"}

trace.set_tracer_provider(TracerProvider())
tracer = trace.get_tracer(__name__)
span_processor = BatchExportSpanProcessor(
    AzureMonitorTraceExporter.from_connection_string(
        os.environ["APPLICATIONINSIGHTS_CONNECTION_STRING "]
    )
)
trace.get_tracer_provider().add_span_processor(span_processor)

FastAPIInstrumentor.instrument_app(app)

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)`

On installing the required packages, there would be a dependency conflict as illustrated in the screenshots in below sections.

**Expected behavior**
Ability to gather telemetry from FastAPI(ASGI) apps; on the lines of this Flask sample posted in this repo
[Flask](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/monitor/azure-monitor-opentelemetry-exporter/samples/traces/sample_server.py)

**Screenshots**
Listing the dependencies version for azure-monitor-opentelemetry-exporter (with allow pre-releases):
![image](https://user-images.githubusercontent.com/43657587/113760475-e6da4300-96db-11eb-9dd9-c81b629ac034.png)

Dependency resolver error when trying to add opentelemetry-instrumentation package:
![image](https://user-images.githubusercontent.com/43657587/113760775-42a4cc00-96dc-11eb-94b7-3df3a85f94ec.png)

**Additional context**
In an attempt to track incoming requests, we followed this article
[OpenCensus FastAPI implementation](https://docs.microsoft.com/en-us/azure/azure-monitor/app/opencensus-python-request#tracking-fastapi-applications)
Unfortunately this instrumentation is leading to OOM issues in our Kubernetes deployment and ended up pivoting to open telemetry.

**Open telemetry version: opentelemetry-instrumentation-fastapi 0.19b0**

**Comment by lzchen** [comment]
The current published package of [azure-monitor-opentelemetry-exporter](https://pypi.org/project/azure-monitor-opentelemetry-exporter/) takes a dependency on `openetelemetry-api` package version `0.17b0`. We have recently release `opentelemetry-api` and version `1.0.0`, which [opentelemetry-instrumentation-fastapi](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/metrics/instrumentation/opentelemetry-instrumentation-fastapi/setup.cfg#L41) takes a dependency on. You have version `0.17b0`, which is resulting in a conflict. The PR to update `azure-monitor-opentelemetry-exporter` to use the `1.0.0` versions of the `opentelemetry-api` package has already been [merged](https://github.com/Azure/azure-sdk-for-python/pull/17747), we are just waiting on the release. The release will happen hopefully by the end of this week, so this should solve your issue. @rakshith91

**Comment by gautam-ergo** [comment]
Awesome, thanks @lzchen for the prompt response. Will check back and report on this issue post release.

**Comment by rakshith91** [comment]
Hi @gautam-ergo ,
thanks for your patience! azure-monitor-opentelemetry-exporter 1.0.0b4 is now released and you must be able to fix this error by updating your package.

Please let us know if there are any other concerns

**Comment by gautam-ergo** [comment]
Hey @rakshith91 awesome stuff.

Dependency conflict has been resolved and the packages now co-exist. 

however, just one clarification required, in the [flask sample](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/monitor/azure-monitor-opentelemetry-exporter/samples/traces/sample_server.py) posted in this repo there is an import pointing to,

**_from opentelemetry.sdk.trace.export import BatchExportSpanProcessor_**

But Cannot find reference 'BatchExportSpanProcessor' in open telemetry's ['__init__.py'](https://github.com/open-telemetry/opentelemetry-python/blob/b9c0b6f8827895e23eeff1f253f4c502154d1873/opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py)

Not sure if this is intended. Kindly advise.

Please feel free to close this issue if you feel the highlighted point warrants another issue for tracking.

Thanks again.

**Comment by rakshith91** [comment]
@gautam-ergo  
try `from opentelemetry.sdk.trace.export import BatchSpanProcessor`  - the sample has been updated too

**Comment by gautam-ergo** [comment]
Thanks @rakshith91. Works as expected.

One last question, we have multiple POST paths in our FastAPI app and how do you suggest to record these request URL's in App Insights performance monitoring.

**Comment by lzchen** [comment]
@gautam-ergo 
What do you see in your App map currently?

**Comment by gautam-ergo** [comment]
@lzchen 

Hope this helps. 

Application map:
![image](https://user-images.githubusercontent.com/43657587/113919860-ad700900-97a9-11eb-88c3-701aa913c60f.png)

Performance tab :
![image](https://user-images.githubusercontent.com/43657587/113919899-b95bcb00-97a9-11eb-82f1-20d29a23f1aa.png)

Swagger :
![image](https://user-images.githubusercontent.com/43657587/113920211-19eb0800-97aa-11eb-9a4c-08310927011d.png)

**Comment by lzchen** [comment]
@gautam-ergo 
> how do you suggest to record these request URL's in App Insights performance monitoring.

Not sure how to answer this without more context. What are you trying to achieve in app insights? Do you need your app map to look a certain way?

**Comment by gautam-ergo** [comment]
@lzchen 

Sorry about the lack of info. 
Essentially looking for a way to implement an APM for one of our python apps using App Insights.

- A way to set up the cloud role name for python app (to avoid _unknown_service_ because of a lack of cloud role name)
- Capturing the exceptions occurring in the POST request(s) so that it appears in the _Failures_ tab
- Set up Availability/Failed Request graph in Application Insights Dashboard (able to see server requests and server response time successfully)

![image](https://user-images.githubusercontent.com/43657587/113941454-bc17e980-97c4-11eb-8ffb-8d98a5e59900.png)

Currently using opentelemetry's azure monitor package with fastapi instrumentation

**Comment by lzchen** [comment]
@gautam-ergo 
> A way to set up the cloud role name for python app (to avoid unknown_service because of a lack of cloud role name)
( Documentation is clear on adding cloud role name via a callback in case of OpenCensus, does the same apply for
OpenTelemetry as well )

Currently, the Opentelemetry exporters do not support telemetry processors anymore. The recommended way to modify the cloud role name is through the [Resource](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/resources/__init__.py#L136) API. You can see the [logic](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/export/trace/_exporter.py#L98) for how cloud role name is populated based off the attributes in the provided `Resource`. Please remember that the OpenTelemetry azure monitor exporter is still in BETA, and these features are subject to change.

> Capturing the exceptions occurring in the POST request(s) so that it appears in the Failures tab

Currently not supported. Will be available once exporter goes GA.

> Set up Availability/Failed Request graph in Application Insights Dashboard (able to see server requests and server response time successfully)

Failed requests will be populated once the above point is implemented. As for `availability`, I do not believe any of our current SDKs sends this type of metric.

**Comment by gautam-ergo** [comment]
Thank you so much @lzchen for the info, it helped a great deal.
Was able to use the Resource API like you mentioned and successfully set the cloud role name.

Final piece of the puzzle, 

- I can see **_Dependency_** in Transaction search, trying to understand the context surrounding this. Are the dependency logs expected and is there a way to not emit them
- Response code is displaying 0, how do I capture the response code for the requests.

Attached a screengrab from the transaction search page for your reference.

![image](https://user-images.githubusercontent.com/43657587/114049680-4443d080-9851-11eb-87b5-5048e7875985.png)

Thanks again.

**Comment by lzchen** [comment]
> I can see Dependency in Transaction search, trying to understand the context surrounding this. Are the dependency logs expected and is there a way to not emit them

Are you instrumented with `RequestsIntrumentor` as well?

> Response code is displaying 0, how do I capture the response code for the requests.

Interesting, the response code should reflect the status code of the request. Can you print out the span that was actually used to track this request? You can use the [console](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py#L379) exporter to do this.

**Comment by gautam-ergo** [comment]
@lzchen 

Yes, I am using the RequestsIntrumentor as well. Is this redundant when using the FastAPI instrumentor.

Able to see the response code in attributes under http.status code

![image](https://user-images.githubusercontent.com/43657587/114062796-88d56900-985d-11eb-9a86-bf63c858ffe9.png)

Additionally,
Attached 3 json files,

[json.zip](https://github.com/Azure/azure-sdk-for-python/files/6280050/json.zip)

one for successful - 201 request
one for failed - 422 request
one on exit - Didn't realize there was an error during exiting. Attached screenshot below as well as a json in the zip file

![image](https://user-images.githubusercontent.com/43657587/114062043-aeae3e00-985c-11eb-91ee-8dae718dd5c8.png)

Also is this an error message >> **_description should only be set when status_code is set to StatusCode.ERROR_**

Please let me know if you need additional info.

Thanks.

**Comment by lzchen** [comment]
@gautam-ergo 
> Yes, I am using the RequestsIntrumentor as well. Is this redundant when using the FastAPI instrumentor.

The `RequestsInstrumentor` will track telemetry when you use the [requests](https://pypi.org/project/requests/) library, so it has nothing to do with FastAPI. The reason you are seeing the dependency telemetry is because our EXPORTER itself is using the `requests` library to make calls to our backend to actually send your telemetry. This is actually a bug on our end, we don't want to be tracking telemetry sent from our exporter, I've already created a bug [here](https://github.com/Azure/azure-sdk-for-python/issues/17904). For now, if you do not want to see these dependencies in app insights, I would simply remove `RequestsInstrumentor` if you do not actually need to track these.

> Attached 3 json files,

Looking at the output, it seems like the spans are missing some attributes. This is probably an issue in the `opentelemetry-instrumentation-asgi`, could you file an issue [here](https://github.com/open-telemetry/opentelemetry-python-contrib) for this problem specifically? Basically, the exporter sees if `http.status_code` is populated in the span attributes to populate `response_Code` in app insights. Since the attribute is not there, this is why you are seeing 0s. Normally, it is up to the instrumentation to populate this.

>  description should only be set when status_code is set to StatusCode.ERROR

This is a bug that is already fixed but just hasn't been released yet.

**Comment by gautam-ergo** [comment]
Thanks @lzchen 
Issue created and tagged and a minor update to the doc as well is required, 

**_from opentelemetry.sdk.trace.export import SimpleExportSpanProcessor_**

[SimpleExportSpanProcessor](https://docs.microsoft.com/en-us/python/api/overview/azure/core-tracing-opentelemetry-readme?view=azure-python-preview#examples)
Used SimpleSpanProcessor instead of SimpleExportSpanProcessor based on previous directions.

Also, knowing its an different issue, would you be able to suggest any config tweaks to
[OpenCensus FastAPI implementation](https://docs.microsoft.com/en-us/azure/azure-monitor/app/opencensus-python-request#tracking-fastapi-applications) to avoid the OOM issue in general.

**Comment by lzchen** [comment]
@rakshith91 
> from opentelemetry.sdk.trace.export import SimpleExportSpanProcessor

SimpleExportSpanProcessor
Used SimpleSpanProcessor instead of SimpleExportSpanProcessor based on previous directions.

Docs for azure core tracing need updating.

> Also, knowing its an different issue, would you be able to suggest any config tweaks to
OpenCensus FastAPI implementation to avoid the OOM issue in general.

Difficult to say what the culprit is in terms of what is causing the OOM. I would suggest removing components one by one until the source of what is not being garbage collected or cleaned up properly can be found. Also, implementing an actual FastAPI integration might be a solution as well.

**Comment by rakshith91** [comment]
@gautam-ergo 

The README fix would be tracked in a different issue.
Are there any other concerns that you have? If not, feel free to close the issue.
Thanks for posting this again.

**Comment by gautam-ergo** [comment]
@rakshith91 

Thanks for all the help here.

Final clarification, is there a way to add custom dimensions on the dependency telemetry similar to what has been showcased in [custom dimensions](https://docs.microsoft.com/en-us/azure/azure-monitor/app/opencensus-python#logs)

@lzchen mentioned
> The reason you are seeing the dependency telemetry is because our EXPORTER itself is using the requests library to make calls to our backend to actually send your telemetry.

**Comment by lzchen** [comment]
@gautam-ergo 
> Final clarification, is there a way to add custom dimensions on the dependency telemetry similar to what has been showcased in custom dimensions

If you add attributes to the span, those attributes will become custom dimensions of the dependency telemetry that it creates. You can use the [on_end](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py#L198) method of the `BatchSpanProcessor`, which is a callback that is called right before each span ends. Currently there is no way to modify a `ReadableSpan` so you would have to convert the `ReadableSpan` to an actual `Span`, modify the attributes, and then convert it back to a `ReadableSpan`, OR the more hacky way, simply access the protected [variable](https://github.com/open-telemetry/opentelemetry-python/blob/b73d8009904fc1693b2ba3a1e0656af376f859cf/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L424) of the `ReadableSpan` directly.

> The reason you are seeing the dependency telemetry is because our EXPORTER itself is using the requests library to make calls to our backend to actually send your telemetry.

If you don't need to use `RequestsInstrumentor()` to track calls from the `requests` library, simply remove it. If you do, you will still see those dependency telemetry appear in your app insights instance until the [bug](https://github.com/Azure/azure-sdk-for-python/issues/17904) is fixed.

**Comment by gautam-ergo** [comment]
Thanks @rakshith91 @lzchen  for all the help.

Closing the issue.

P.S. unlike the OpenCensus implementation, did not encounter OOM issue whilst using the opentelemetry azure monitor with fastapi instrumentation package. Thanks again for all this awesome work.
